src/main/java/com/salesforce/cdp/queryservice/util/TokenHelper.java

```diff
@@ -134,6 +144,57 @@ private static Token retrieveTokenWithPasswordGrant(Properties properties, OkHtt
         }
     }
 
+    private static Token retrieveTokenWithJWTBearerGrant(Properties properties, OkHttpClient client) throws TokenException {
+        if (properties.getProperty(Constants.USER_NAME) == null || properties.getProperty(Constants.USER_NAME).isEmpty()) {
+            throw new TokenException("Username cannot be null/empty for key-pair authentication");
+        }
+
+        if (properties.getProperty(Constants.PRIVATE_KEY) == null || properties.getProperty(Constants.PRIVATE_KEY).isEmpty()) {
+            throw new TokenException("Private key cannot be null/empty for key-pair authentication");
+        }
+
+        if (properties.getProperty(Constants.CLIENT_ID) == null || properties.getProperty(Constants.CLIENT_ID).isEmpty()) {
+            throw new TokenException("Client Id cannot be null/empty for key-pair authentication");
+        }
+
+        String token_url = properties.getProperty(Constants.LOGIN_URL) + Constants.CORE_TOKEN_URL;
+        CoreTokenRenewResponse coreTokenRenewResponse = null;
+        try {
+            String audience = getAudienceForJWTAssertion(properties.getProperty(Constants.LOGIN_URL));
+
+            // And handle the initial login *without* immutables. This ensures that nothing
+            // is allocated in memory that cannot be cleared on demand, and thus we aren't
+            // at the garbage collectors mercy.
+            Response response = key_pair_auth_login(Constants.TOKEN_GRANT_TYPE_JWT_BEARER,
+                    properties.getProperty(Constants.CLIENT_ID),
+                    properties.getProperty(Constants.USER_NAME),
+                    properties.getProperty(Constants.PRIVATE_KEY),
+                    audience, token_url, client);
+
+            // And exchange the Key/Pair flow authtoken for a scoped bearer token.
+            coreTokenRenewResponse = HttpHelper.handleSuccessResponse(response, CoreTokenRenewResponse.class, false);
+            return exchangeToken(coreTokenRenewResponse.getInstance_url(), coreTokenRenewResponse.getAccess_token(), client);
+        } catch(SQLException sqlException) {
+          log.error("Caught exception while setting audience for JWT assertion", sqlException);
+          throw new TokenException(TOKEN_FETCH_FAILURE, sqlException);
+        } catch (IOException e) {
+            log.error("Caught exception while retrieving the token", e);
+            throw new TokenException(TOKEN_FETCH_FAILURE, e);
+        }
+    }
+
+    private static String getAudienceForJWTAssertion(String serviceRootUrl) throws SQLException {
+        String serverUrl = serviceRootUrl.toLowerCase();
+        //STMPA and STMPB envs are ignored as they are not used anymore.
+        if (serverUrl.endsWith(Constants.NA45_SERVER_URL) || serverUrl.endsWith(Constants.NA46_SERVER_URL)) {
```

Comment:
Please confirm if Aud for NA45 is same as NA46

Reply:
Aud for NA45, NA46 and dev envs is same.